### PR TITLE
[FIX] tools: support malformed more malformed pdf

### PR DESCRIPTION
**Pull request by william-andre**
We do not want to fail ever if loading the attachment fails.
We still log a warning in every case though.

Some malformed PDFs lead to seeking the wrong bytes while looking for
some parts, raising a ValueError

See https://github.com/mstamy2/PyPDF2/issues/183

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by smetl** [comment]
@william-andre Plz rebase! @robodoo delegate+

**Comment by william-andre** [comment]
@smetl still ok with this?

File: odoo/tools/pdf.py
Comment (xmo-odoo) [inline odoo/tools/pdf.py:41]: `tok.isdigit()`  is probably cheaper, and certainly clearer. Or `tok in string.digits` if we specifically only want to include western arabic numerals (0-10).

**Comment by xmo-odoo** [comment]
The logic seems... odd. We're just skipping over non-digits until we find whitespace?

File: odoo/tools/pdf.py
Comment (william-andre) [reply odoo/tools/pdf.py]: `tok.isdecimal()` then because what we want to do is form numbers in base 10
https://docs.python.org/3/library/stdtypes.html#str.isdecimal

**Comment by william-andre** [comment]
> The logic seems... odd. We're just skipping over non-digits until we find whitespace?

We read only digits until we find a whitespace. We could do this, or simply catch `ValueError` when doing a `get` on a `PdfObject`.
We are having the error because some PDFs are malformed (I trust the maintainers of PyPDF2 on this one). Anyways, it is already reported in PyPDF2, so the only thing we can do is monkey patch it here waiting for a fix (since 2014...)

**Comment by william-andre** [comment]
The traceback I got without this fix

```
2021-03-05 09:36:40,776 45194 ERROR 14.0-edi-pdf-attachment-wan odoo.http: Exception during JSON request handling.
Traceback (most recent call last):
  File "/home/william/git/odoo/odoo/odoo/addons/base/models/ir_http.py", line 237, in _dispatch
    result = request.dispatch()
  File "/home/william/git/odoo/odoo/odoo/http.py", line 683, in dispatch
    result = self._call_function(**self.params)
  File "/home/william/git/odoo/odoo/odoo/http.py", line 359, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/home/william/git/odoo/odoo/odoo/service/model.py", line 94, in wrapper
    return f(dbname, *args, **kwargs)
  File "/home/william/git/odoo/odoo/odoo/http.py", line 347, in checked_call
    result = self.endpoint(*a, **kw)
  File "/home/william/git/odoo/odoo/odoo/http.py", line 912, in __call__
    return self.method(*args, **kw)
  File "/home/william/git/odoo/odoo/odoo/http.py", line 531, in response_wrap
    response = f(*args, **kw)
  File "/home/william/git/odoo/odoo/addons/web/controllers/main.py", line 1377, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/home/william/git/odoo/odoo/addons/web/controllers/main.py", line 1369, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/home/william/git/odoo/odoo/odoo/api.py", line 396, in call_kw
    result = _call_kw_multi(method, model, args, kwargs)
  File "/home/william/git/odoo/odoo/odoo/api.py", line 383, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/home/william/git/odoo/odoo/addons/account/models/account_journal.py", line 598, in create_invoice_from_attachment
    invoice = decoder[1](attachment)
  File "/home/william/git/odoo/odoo/addons/account_edi/models/account_edi_format.py", line 367, in _create_invoice_from_attachment
    for file_data in self._decode_attachment(attachment):
  File "/home/william/git/odoo/odoo/addons/account_edi/models/account_edi_format.py", line 355, in _decode_attachment
    to_process.extend(self._decode_pdf(attachment.name, content))
  File "/home/william/git/odoo/odoo/addons/account_edi/models/account_edi_format.py", line 325, in _decode_pdf
    for xml_name, content in pdf_reader.getAttachments():
  File "/home/william/git/odoo/odoo/odoo/tools/pdf.py", line 130, in getAttachments
    if not self.trailer["/Root"].get("/Names", {}).get("/EmbeddedFiles", {}).get("/Names"):
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/generic.py", line 516, in __getitem__
    return dict.__getitem__(self, key).getObject()
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/generic.py", line 178, in getObject
    return self.pdf.getObject(self).getObject()
  File "/home/william/git/odoo/odoo/odoo/tools/pdf.py", line 36, in getObject
    return originalGetObject(self, indirectReference)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/pdf.py", line 1599, in getObject
    idnum, generation = self.readObjectHeader(self.stream)
  File "/home/william/git/odoo/odoo/odoo/tools/pdf.py", line 29, in readObjectHeader
    return originalReadObjectHeader(self, stream)
  File "/usr/local/lib/python3.8/dist-packages/PyPDF2/pdf.py", line 1667, in readObjectHeader
    return int(idnum), int(generation)
Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/william/git/odoo/odoo/odoo/http.py", line 639, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/home/william/git/odoo/odoo/odoo/http.py", line 315, in _handle_exception
    raise exception.with_traceback(None) from new_cause
ValueError: invalid literal for int() with base 10: b'ndobj'
```

I just printed some values on the stream we got in `readUntilWhitespace`
```python
>>> string = stream.read()
b'ndobj\n2 0 obj\n<< /Type /Catalog\n/Pages 1 0 R\n >>\nendobj\n3 0 obj\n<< \n/Font << /F1 9 0 R /F3 10 0 R >> \n/ProcSet [ /PDF /ImageC /Text ] /XObject <</Im1 6 0 R \n/Im2 7 0 R \n >>\n>> \nendobj\nxref\n0 11\n0000000000 65535 f \n0000092566 00000 n \n0000092624 00000 n \n0000092674 00000 n \n0000000015 00000 n \n0000000071 00000 n \n0000004250 00000 n \n0000008953 00000 n \n0000092240 00000 n \n0000092346 00000 n \n0000092453 00000 n \ntrailer\n<<\n/Size 11\n/Root 2 0 R\n/Info 4 0 R\n>>\nstartxref\n92802\n%%EOF\n'
>>> stream = io.BytesIO(string)
```

**Comment by odony** [comment]
> We are having the error because some PDFs are malformed (I trust the maintainers of PyPDF2 on this one). Anyways, it is already reported in PyPDF2, so the only thing we can do is monkey patch it here waiting for a fix (since 2014...)

There won't be any fix, because the real issue is that these PDF files have an invalid `xref` table. This means the parser ends up in the middle of a random PDF object instead of the target object header. You can try to pretend like everything is normal and proceed with accepting random input as valid, but that's not gonna end well. It will just randomly blow up later with a different exception, or simply fail to parse any sense out of the PDF, and think it's an empty file.

Imagine you have a database where every SQL query returns random records instead of the ones you asked for, including record from other tables. What's your strategy to retrieve the data you're looking for?

I see only 2 options here:

  1. Implement a full-fledged **auto-repair** for the xref table. That's what most PDF viewers do, and that explains why they can display files that pyPDF2 can't parse. By the way you can do it manually, e.g. with `mutool clean`, and that will make the file readable by pyPDF2.
  2. Or consider the PDF corrupt as soon as you encounter any parsing exception, and bail out with a warning.

The implementation of option 1 in pyPDF2 is *extremely unlikely* given the current state of the project, so don't hold your breath. We could do it ourselves, but that sounds out of scope for your current PR in 14.

So I suggest we go for option 2. Don't make random monkey-patches trying hide/fix the unfixable, just catch *any exception* when calling pyPDF2 methods. `getAttachments()` is a good candidate for that, because it goes deep in the PDF structure, and definitely won't work if the xref table is bad (plus we already override it to support multiple attachments)


----

PS: If you're curious about PDF internals, you should have a look at those [brilliant slides](https://speakerdeck.com/ange/lets-write-a-pdf-file). You will be hand-crafting PDF files in minutes, and you'll understand everything about this issue much better. That's what I used to generate the example `minimal.pdf` file that we use for [our PDF tests](https://github.com/odoo/odoo/blob/master/odoo/addons/base/tests/test_pdf.py#L15).

PS2: If you want to reproduce PDF parsing problems it's very easy, edit `minimal.pdf` and modify randomly the 10-digits numbers (which represent offsets inside the file) at the end of the file. e.g. divide them by 2, or subtract 16 to each of them, because why not?
screenshot-github com-2021 03 05-18_33_18 

PS3: The PDF file that triggered this problem (courtesy of @avw-odoo) was produced by Apache FOP v0.20.5, a super old beta version from the ice age. I doubt that version supported embedded attachments, and even if it did, that would not be a terrible loss if we can't extract them, frankly.

**Comment by smetl** [comment]
Are we good on this one?

**Review by odony — APPROVED** [review]
Thanks! The commit headline is a bit weird, but I get we get the point ;-)

@robodoo r+

**Pull request closed without merging**
